### PR TITLE
hoon: add pin and awl arms to test and trim subaxes

### PR DESCRIPTION
**Pull request by eamsden**
`+pin`: test if `b` is a sub-axis of `a`
`+awl`: give the remainder axis of `b` under `a` (crash if `b` isn't a sub-axis of `a`)

### CODE REVIEW

File: pkg/arvo/sys/hoon.hoon [hunk 1]
```diff
@@ -226,6 +226,33 @@
...
+  ::
+  ::  checks whether {b} addresses a subtree of the tree addressed by {a}.
+  |=  [a=@ b=@]
+  ?<  =(0 a)
+  ?<  =(0 b)
+  |-  ^-  ?
+  ?:  =(a 1)  %&
+  ?:  =((cap a) (cap b))
+    $(a (mas a), b (mas b))
+  %|
```
Comment (joemfb) [inline pkg/arvo/sys/hoon.hoon:241]: To ensure that the last sub-hoon is the largest, can you switch this to a `?.`? Also, this needs a single `::` line between the `+pin` and `+awl` arms.

File: pkg/arvo/sys/hoon.hoon [hunk 2]
```diff
@@ -226,6 +226,33 @@
...
+  ::  computes the remainder of axis {b} when navigating to {a}.
+  ::  (crashes if not `(pin a b)`)
+  |=  [a=@ b=@]
+  ?<  =(0 a)
+  ?<  =(0 b)
+  |-  ^-  @
+  ?:  =(a 1)  b
+  ?:  =((cap a) (cap b))
+    $(a (mas a), b (mas b))
+  !!
```
Comment (joemfb) [inline pkg/arvo/sys/hoon.hoon:254]: This should be a `?>`.

File: pkg/arvo/sys/hoon.hoon [hunk 3]
```diff
@@ -226,6 +226,33 @@
...
+  ?<  =(0 a)
+  ?<  =(0 b)
+  |-  ^-  ?
+  ?:  =(a 1)  %&
+  ?:  =((cap a) (cap b))
+    $(a (mas a), b (mas b))
+  %|
+++  awl
+  ~/  %awl
+  ::    axis after axis
```
Comment (joemfb) [inline pkg/arvo/sys/hoon.hoon:244]: The should also be a blank `::` line after this "doccord summary".

**Review by joemfb — CHANGES_REQUESTED** [review]

**Comment by eamsden** [comment]
just realized `+awl` isn't cvc... don't have any replacement name ideas

File: pkg/arvo/sys/hoon.hoon [hunk 4]
```diff
@@ -226,6 +226,33 @@
...
+++  pin
+  ~/  %pin
+  ::    sub axis test
+  ::
+  ::  checks whether {b} addresses a subtree of the tree addressed by {a}.
+  |=  [a=@ b=@]
+  ?<  =(0 a)
+  ?<  =(0 b)
+  |-  ^-  ?
+  ?:  =(a 1)  %&
```
Comment (joemfb) [inline pkg/arvo/sys/hoon.hoon:238]: I think `+pin` should be total over non-zero atoms. For that to be the case, this should be

```
?:  =(a 1)  &
?:  =(b 1)  |
```

**Review by joemfb — CHANGES_REQUESTED** [review]

**Review by joemfb — APPROVED** [review]
These look right to me. They're not urgent, and probably won't go into the next release. They're not particularly useful without jets (which I'd be happy to write). Let's defer to @belisarius222 as to when/where to merge.

this will do

**Comment by joemfb** [comment]
@eamsden do you expect to always use these gates together? would it be better to combine them into a single gate producing `(unit axis)`?

**Comment by eamsden** [comment]
> @eamsden do you expect to always use these gates together? would it be better to combine them into a single gate producing `(unit axis)`?

No I expect to use `+pin` by itself quite frequently. `+hub` was frankly only included because it seemed a shame not to when you're already doing all the math.

**Review by ashelkovnykov — APPROVED** [review]
Just making sure I understand correctly: `hub` is `mas of b relative to a`, correct?

**Comment by eamsden** [comment]
> Just making sure I understand correctly: `hub` is `mas of b relative to a`, correct?

Roughly yes. Instead of taking off 1 bit it takes off however many bits are in `a` (not counting the MSB).

Notably though it crashes if `b` is not a subaxis of `a`.

**Comment by eamsden** [comment]
@jalehman note that these likely depend on Joe's jets in urbit/vere#397

**Pull request merged**